File: Week3_Arrays/coin-change.java
import java.util.*;

class Solution {
    public int coinChange(int[] coins, int amount) {
        int max = amount + 1;
        if (amount == 0) {
            return 0;
        }

        int[] dp = new int[amount + 1];
        Arrays.fill(dp, max);
        dp[0] = 0;
        Arrays.sort(coins);
        for (int coin : coins) {
            for (int idx = coin; idx < dp.length; idx++) {
                dp[idx] = Math.min(dp[idx], dp[idx - coin] + 1);
            }
        }

        if (dp[amount] == max) {
            return -1;
        }
        return dp[amount];
    }
}

File: Week3_Arrays/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0
        
        dp = [float('inf') for _ in range(amount + 1)]
        dp[0] = 0
        coins.sort()
        for coin in coins:
            for i in range(coin, amount + 1):
                dp[i] = min(dp[i], dp[i - coin] + 1)
        
        return dp[amount] if dp[amount] != float('inf') else -1